app/controllers/job_invocations_controller.rb

```diff
@@ -2,16 +2,28 @@ class JobInvocationsController < ApplicationController
   include Foreman::Controller::AutoCompleteSearch
 
   def new
-    @composer = JobInvocationComposer.new(JobInvocation.new,
-                                          :host_ids => params[:host_ids],
-                                          :targeting => {
-                                            :targeting_type => Targeting::STATIC_TYPE,
-                                            :bookmark_id => params[:bookmark_id]
-                                          })
+    @composer = JobInvocationComposer.new.compose_from_params(
+      :host_ids => params[:host_ids],
+      :targeting => {
+        :targeting_type => Targeting::STATIC_TYPE,
+        :bookmark_id => params[:bookmark_id]
+      })
+  end
+
+  def rerun
+    job_invocation = resource_base.find(params[:id])
+    @composer = JobInvocationComposer.new.compose_from_invocation(job_invocation)
+
+    if params[:failed_only]
```

Comment:
Ok, let it here for now: if the need for sharing this logic appears, we can move the code on more sharable place